app/uk/gov/hmrc/gform/controllers/AuthenticatedRequestActions.scala

```diff
@@ -195,9 +197,17 @@ class AuthenticatedRequestActions(
     maybeAccessCode: Option[AccessCode],
     formTemplate: FormTemplate)(retrievals: MaterialisedRetrievals)(implicit hc: HeaderCarrier): Future[Result] =
     for {
-      form        <- gformConnector.getForm(FormId(retrievals, formTemplate._id, maybeAccessCode))
-      obligations <- obligationService.lookupObligationsMultiple(formTemplate)
-      result      <- f(AuthCacheWithForm(retrievals, form, formTemplate, obligations))
+      form    <- gformConnector.getForm(FormId(retrievals, formTemplate._id, maybeAccessCode))
+      newForm <- obligationService.lookupIfPossible(form, formTemplate, authService, retrievals)
+      update <- obligationService.updateObligations(
```

Comment:
Why are we calling `updateObligations` here? It seems like completely wrong thing to do. Why aren't we doing this logic in `FormController`?

Reply:
@kylemchownhmrc , updateObligations is the last thing called before f and calling it in AuthenticatedRequests is not ready early appropriate.  That function f has been passed in (via async) from the controller.

So it would be more appropriate to move the call out to the beginning of the functions for, e.g. uk.gov.hmrc.gform.gform.FormController#form and uk.gov.hmrc.gform.gform.SummaryController#submit .  

Note that in some of the uses, e.g. DeclarationController and AcknowledgementController it is actually inappropriate to update the userData, because the form has already been finally validated or indeed submitted.  They can actually call uk.gov.hmrc.gform.controllers.AuthenticatedRequestActions#asyncWithoutObligations instead.  You can almost get away with asyncWithoutObligations in most or all cases except for the two above, so leaving just two calls to updateObligations, one for form sections and the second for summary, where final validation occurs.

You will just need to add form to uk.gov.hmrc.gform.controllers.AuthCacheWithForm , so that the functions called can call updateObligations with form as well as newForm.

Do you see?